packages/broker/src/plugins/legacyWebsocket/WebsocketServer.ts

```diff
@@ -87,179 +103,146 @@ export class WebsocketServer extends EventEmitter {
-                }
-            },
-            close: (ws: uWS.WebSocket, _code: number, _message: ArrayBuffer) => {
-                const connection = this.connections.get(ws.connectionId)
-                if (connection) {
-                    logger.trace('closing socket "%s" on streams "%o"', connection.id, connection.streamsAsString())
-                    this._removeConnection(connection)
+                    connection.send(new ControlLayer.ErrorResponse({
+                        version: request.version,
+                        requestId: request.requestId,
+                        errorMessage: err.message || err,
+                        errorCode: err.errorCode || 'ERROR_WHILE_HANDLING_REQUEST',
+                    }))
                 }
-            },
-            pong: (ws: uWS.WebSocket) => {
-                const connection = this.connections.get(ws.connectionId)
-
-                if (connection) {
-                    logger.trace(`received from ${connection.id} "pong" frame`)
-                    // @ts-expect-error
-                    connection.respondedPong = true
-                }
-            }
+            })
+            connection.once('close', () => {
+                this.removeConnection(connection)
+            })
+            this.connections.add(connection)
         })
-    }
 
-    static validateProtocolVersions(controlLayerVersion: number|undefined, messageLayerVersion: number|undefined) {
-        if (controlLayerVersion === undefined || messageLayerVersion === undefined) {
-            throw new Error('Missing version negotiation! Must give controlLayerVersion and messageLayerVersion as query parameters!')
-        }
+        this.wss.on('error', (err) => {
+            logger.error(`websocket server error: %s`, err)
+        })
 
-        // Validate that the requested versions are supported
-        if (ControlLayer.ControlMessage.getSupportedVersions().indexOf(controlLayerVersion) < 0) {
-            throw new Errors.UnsupportedVersionError(controlLayerVersion, `Supported ControlLayer versions: ${
-                JSON.stringify(ControlLayer.ControlMessage.getSupportedVersions())
-            }. Are you using an outdated library?`)
-        }
+        this.pingInterval = setInterval(() => {
+            this.pingConnections()
+        }, this.pingIntervalInMs)
+    }
 
-        if (MessageLayer.StreamMessage.getSupportedVersions().indexOf(messageLayerVersion) < 0) {
-            throw new Errors.UnsupportedVersionError(messageLayerVersion, `Supported MessageLayer versions: ${
-                JSON.stringify(MessageLayer.StreamMessage.getSupportedVersions())
-            }. Are you using an outdated library?`)
-        }
+    async close(): Promise<unknown> {
+        clearInterval(this.pingInterval)
+        this.requestHandler.close()
+        this.connections.forEach((connection: Connection) => connection.close())
+        return new Promise((resolve, reject) => {
```

Comment:
Use `promisify`? E.g.
 
```
await util.promisify((cb: any) => this.wss.close(cb))()
```

Reply:
Any thoughts on how to do this with the current code where there are two internal closes?